From 3dfcd2415a999df7ea6e54db06d13b5e2b8deb0f Mon Sep 17 00:00:00 2001
From: Meron Cohen <mecohen@habana.ai>
Date: Sat, 07 Sep 2024 11:07:26 +0300
Subject: [PATCH] [SW-199639] g3: common: handle device id

Now that both OAM and PCI card type exist, the card type is detected
and the proper control bit in PIF arc control register is set.
To get the device ID, there is to read this bit.

Change-Id: I08dd554f03198c9059bbd0d637272e99700e5aac
Signed-off-by: Meron Cohen <mecohen@habana.ai>
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 68d80bc..80a6b77 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -1125,6 +1125,7 @@
 	if (preboot_is_bis_compliant_mode())
 		pif_set_bis_compliant_device();
 
+	preboot_device_id_init();
 	ppboot_load_and_run_pif_arc(&load_data);
 }
 
diff --git a/arch/gaudi3/common/common.c b/arch/gaudi3/common/common.c
index 2bd33b7..b13806c 100644
--- a/arch/gaudi3/common/common.c
+++ b/arch/gaudi3/common/common.c
@@ -20,6 +20,9 @@
 #include <bitops.h>
 #include <gpio_regs.h>
 #include <address_space.h>
+#include <common/pci_ids.h>
+#include <pif_arc_special_regs.h>
+#include <hl_pif_arc_if.h>
 
 #define GPIOS_PER_BANK			16
 #define FORCE_RECOVERY_MASK	BIT(2)
@@ -146,3 +149,31 @@
 {
 	return D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_PAD_SEL_18);
 }
+
+#define GPIO_CARD_TYPE		23
+#define GPIO_CARD_TYPE_BIT	BIT(GPIO_CARD_TYPE % 16)
+
+static bool is_card_type_pcie(void)
+{
+	uint32_t val;
+	uint32_t is_pcie;
+
+	val = regmap_readl(D0_PSOC_GPIO1_REG(mmGPIO_DATA));
+
+	is_pcie = val & GPIO_CARD_TYPE_BIT;
+
+	return !!is_pcie;
+}
+
+void preboot_device_id_init(void)
+{
+	if (is_card_type_pcie())
+		regmap_rmwritel(mmD0_PIF_ARC_CONTROL, 1, PIF_CTRL_PCI_CARD);
+}
+
+u32 preboot_device_id_get(void)
+{
+	u32 arc_ctrl = regmap_readl(mmD0_PIF_ARC_CONTROL);
+
+	return (arc_ctrl & PIF_CTRL_PCI_CARD) ? PCI_IDS_GAUDI3_HL_338 : PCI_IDS_GAUDI3;
+}
